File: NEWCODER/JZ17.cpp
/**
 * 
 * 题目：
 * 输入两棵二叉树A，B，判断B是不是A的子结构。
 * （ps：我们约定空树不是任意一个树的子结构）
 * 
 * 示例：
 * 输入：
 * {8,8,#,9,#,2,#,5},{8,9,#,2}
 * 输出：
 * true
 * 
 * 要求：
 * 时间限制：C/C++ 1秒，其他语言2秒
 * 空间限制：C/C++ 64M，其他语言128M
 */ 

#include<iostream>
#include<vector>

using namespace std;

struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};

bool dfs(TreeNode *r1, TreeNode *r2) {
    if (!r2) return true;
    if (!r1) return false;
    return r1->val==r2->val && dfs(r1->left, r2->left) && dfs(r1->right, r2->right);
}

bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2){
    if (!pRoot1 || !pRoot2) return false;
    return dfs(pRoot1, pRoot2) || HasSubtree(pRoot1->left, pRoot2) || HasSubtree(pRoot1->right, pRoot2);
}
int main(){
    TreeNode* pRoot1 = new TreeNode(8);
    pRoot1->left = new TreeNode(8);
    pRoot1->left->left = new TreeNode(9);
    pRoot1->left->left->left = new TreeNode(2);
    pRoot1->left->left->left->left = new TreeNode(5);

    TreeNode* pRoot2 = new TreeNode(8);
    pRoot2->left = new TreeNode(9);
    pRoot2->left->left = new TreeNode(2);

    cout << HasSubtree(pRoot1,pRoot2) << endl;
    system("pause");
    return 1;
}

File: NEWCODER/JZ17.py
# /**
#  *
#  * 题目：
#  * 输入两棵二叉树A，B，判断B是不是A的子结构。
#  * （ps：我们约定空树不是任意一个树的子结构）
#  *
#  * 示例：
#  * 输入：
#  * {8,8,#,9,#,2,#,5},{8,9,#,2}
#  * 输出：
#  * true
#  *
#  * 要求：
#  * 时间限制：C/C++ 1秒，其他语言2秒
#  * 空间限制：C/C++ 64M，其他语言128M
#  */
# -*- coding:utf-8 -*-

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def HasSubtree(pRoot1, pRoot2):
    # write code here
    result = False
    if pRoot1 and pRoot2:
        if pRoot1.val == pRoot2.val:
            result = same(pRoot1, pRoot2)
        if not result:
            result = HasSubtree(pRoot1.left, pRoot2)
        if not result:
            result = HasSubtree(pRoot1.right, pRoot2)
    return result


def same(p1, p2):
    if not p2:
        return True
    if not p1:
        return False
    return p1.val == p2.val and same(p1.left,p2.left) and same(p1.right,p2.right)


# example
if __name__ == '__main__':
    l1 = TreeNode(8)
    l1.left = TreeNode(8)
    l1.left.left = TreeNode(9)
    l1.left.left.left = TreeNode(2)
    l1.left.left.left.left = TreeNode(5)

    l2 = TreeNode(8)
    l2.left = TreeNode(9)
    l2.left.left = TreeNode(2)

    result = HasSubtree(l1,l2)
    print(result)                       # True



